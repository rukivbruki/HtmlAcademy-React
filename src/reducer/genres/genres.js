import {extend} from "../../utils/extend";

const initialState = {
  genres: [],
  genreFilter: `All genres`,
};

export const ActionType = {
  SET_GENRES: `SET_GENRES`,
  SELECT_GENRE_FILTER: `SELECT_GENRE_FILTER`,
};

export const ActionCreator = {
  setGenres: (list) => ({
    type: ActionType.SET_GENRES,
    payload: list
  }),
  selectGenreFilter: (genre) => ({
    type: ActionType.SELECT_GENRE_FILTER,
    payload: genre
  }),

};

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.SET_GENRES:
      return extend(state, {
        genres: action.payload
      });
    case ActionType.SELECT_GENRE_FILTER:
      return extend(state, {
        genreFilter: action.payload
      });
    default:
      return state;
  }
};

