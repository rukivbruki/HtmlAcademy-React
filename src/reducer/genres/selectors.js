import NameSpace from "../name-space.js";

const NAME_SPACE = NameSpace.GENRES;

export const getGenres = (state) => state[NAME_SPACE].genres;
export const getGenreFilter = (state) => state[NAME_SPACE].genreFilter;
