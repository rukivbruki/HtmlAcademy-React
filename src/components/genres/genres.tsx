// Libs
import * as React from "react";
import {connect} from "react-redux";
// Utils
import {FilmsType} from '../../types';
import {ActionCreator} from "../../reducer/genres/genres.js";
import {getGenreFilter, getGenres} from "../../reducer/genres/selectors";
import {getFilms} from "../../reducer/films/selectors";
// Components
import GenresItem from '../genres-item/genres-item';

const SELECTED_GENRE_CLASS = `catalog__genres-item--active`;

interface Props {
  genres: string[];
  genreFilter?: string;
  films: FilmsType;
  onSelectGenre: (genre: string, films: FilmsType) => void;
}

const Genres = ({genres = [], genreFilter, films, onSelectGenre}: Props) => {
  const handleOnSelectGenre = (e, genre) => {
    e.preventDefault();

    onSelectGenre(genre, films);
  };

  const preparedList = genres.map((genre) => {
    const wrapperHandlerOnSelectGenre = (e) => handleOnSelectGenre(e, genre);
    const activeClass = genreFilter === genre ? SELECTED_GENRE_CLASS : ``;

    return (
      <GenresItem
        key={genre}
        genre={genre}
        onClick={wrapperHandlerOnSelectGenre}
        activeClass={activeClass}/>
    );
  });

  return (
    <ul className="catalog__genres-list">
      {preparedList}
    </ul>
  );
};

Genres.defaultProps = {
  selectedGenre: `All genres`,
  genresList: [],
  onSelectGenre: () => void 0,
  films: []
};

const mapStateToProps = (state) => ({
  genres: getGenres(state).slice(0, 8),
  genreFilter: getGenreFilter(state),
  films: getFilms(state)
});

const mapDispatchToProps = (dispatch) => ({
  onSelectGenre: (genre) => {
    dispatch(ActionCreator.selectGenreFilter(genre));
  }
});

export {Genres};
export default connect(mapStateToProps, mapDispatchToProps)(React.memo(Genres));
