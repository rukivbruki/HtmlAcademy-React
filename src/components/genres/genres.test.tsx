import * as React from "react";
import * as renderer from "react-test-renderer";
import {Genres} from "./genres";

const genres = [`genre1`, `genre2`, `genre3`];

const film = {
  backgroundColor: `red`,
  backgroundImage: `img`,
  description: `desc`,
  director: `dir`,
  genre: `genre`,
  id: 1,
  isFavorite: false,
  name: `film`,
  posterImage: `img`,
  previewImage: `img`,
  previewVideoLink: `video`,
  rating: 1,
  released: 1,
  runTime: 1,
  scoresCount: 1,
  starring: [`test`],
  videoLink: `video`
};

const films = [film];
const genreFilter = `test1`;

it(`Genres renders correctly`, () => {
  const tree = renderer
    .create(
        <Genres
          genres={genres}
          genreFilter={genreFilter}
          films={films}
          onSelectGenre={() => void 0} />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
