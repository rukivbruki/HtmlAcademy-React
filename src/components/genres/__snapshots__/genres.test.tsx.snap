// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Genres renders correctly 1`] = `
<ul
  className="catalog__genres-list"
>
  <li
    className="catalog__genres-item "
  >
    <a
      className="catalog__genres-link"
      href="#"
      onClick={[Function]}
    >
      genre1
    </a>
  </li>
  <li
    className="catalog__genres-item "
  >
    <a
      className="catalog__genres-link"
      href="#"
      onClick={[Function]}
    >
      genre2
    </a>
  </li>
  <li
    className="catalog__genres-item "
  >
    <a
      className="catalog__genres-link"
      href="#"
      onClick={[Function]}
    >
      genre3
    </a>
  </li>
</ul>
`;
